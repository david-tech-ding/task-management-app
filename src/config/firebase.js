import { initializeApp } from "firebase/app";
import { getAuth } from "firebase/auth";

const firebaseConfig = {
  apiKey: "AIzaSyAJv-vJvf585MZ0poOjBWynBBiq9W4fCl4",
  authDomain: "task-management-app-d0b1d.firebaseapp.com",
  projectId: "task-management-app-d0b1d",
  storageBucket: "task-management-app-d0b1d.appspot.com",
  messagingSenderId: "357587538083",
  appId: "1:357587538083:web:2850e62cf8ae51bf87af16",
  measurementId: "G-YB9G43E89B",
};

const app = initializeApp(firebaseConfig);

export const auth = getAuth(app);

export default app;
