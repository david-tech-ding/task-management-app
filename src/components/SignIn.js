import React, { useState, useRef, useEffect, useContext } from "react";
import axios from "axios";
import AuthContext from "../context/AuthProvider";
import { useNavigate } from "react-router-dom";
import { signInWithEmailAndPassword } from "firebase/auth";
import { auth } from "../config/firebase";

const SignIn = ({ onSetLoggedInUser, loggedInUser }) => {
  const [userLogin, setUserLogin] = useState({
    email: "",
    password: "",
  });
  const { setAuth } = useContext(AuthContext);
  const userRef = useRef();
  const errRef = useRef();
  const formRef = useRef();
  const navigate = useNavigate();

  const [errMsg, setErrMsg] = useState("");

  const handleEnterKeyPress = (e) => {
    if (e.keyCode === 13) {
      e.preventDefault();
      handleSignIn(e);
    }
  };

  const addEnterKeyPressListener = () => {
    if (userRef.current) {
      userRef.current.addEventListener("keydown", handleEnterKeyPress);
    }
    if (formRef.current) {
      formRef.current.addEventListener("keydown", handleEnterKeyPress);
    }
  };
  const removeEnterKeyPressListener = () => {
    if (userRef.current) {
      userRef.current.removeEventListener("keydown", handleEnterKeyPress);
    }
    if (formRef.current) {
      formRef.current.removeEventListener("keydown", handleEnterKeyPress);
    }
  };
  useEffect(() => {
    if (userRef.current) {
      userRef.current.focus();
    }
  }, []);

  useEffect(() => {
    setErrMsg("");
  }, [userLogin.email, userLogin.password]);

  const handleChange = (e) => {
    setUserLogin({ ...userLogin, [e.target.name]: e.target.value });
  };

  const handleSignIn = async (e) => {
    e.preventDefault();
    removeEnterKeyPressListener();

    try {
      const userCredential = await signInWithEmailAndPassword(
        auth,
        userLogin.email,
        userLogin.password
      );
      const { user } = userCredential;
      axios
        .get(`/user/username/${user.displayName}`)
        .then((response) => {
          onSetLoggedInUser({
            ...loggedInUser,
            id: response.data[0].id,
            userName: user.displayName,
            firstName: response.data[0].firstName,
          });
        })
        .catch((err) => console.log(err));

      setUserLogin("");

      setAuth({
        user: user.email,
        roles: [],
      });
    } catch (err) {
      const errorCode = err.code;
      if (errorCode === "auth/user-not-found") {
        setErrMsg("User not found");
      } else if (errorCode === "auth/wrong-password") {
        setErrMsg("Wrong password");
      } else {
        setErrMsg("Login Failed");
      }
      errRef.current.focus();
    }

    loggedInUser.firstName ? navigate("/") : navigate("/create-user");

    addEnterKeyPressListener();
  };

  return (
    <div className="auth">
      <p
        ref={errRef}
        className={errMsg ? "errmsg" : "offscreen"}
        aria-live="assertive"
      >
        {errMsg}
      </p>
      <h2 className="auth-subheading">Sign In</h2>
      <form className="auth-form" onSubmit={handleSignIn} ref={formRef}>
        <label htmlFor="sign-in_email-input">
          Email
          <input
            className="auth-form_input"
            onChange={handleChange}
            value={userLogin.email}
            type="email"
            name="email"
            id="sign-in_email-input"
            ref={userRef}
            required
          />
        </label>
        <label htmlFor="sign-in_password-input">
          Password
          <input
            className="auth-form_input"
            onChange={handleChange}
            value={userLogin.password}
            type="password"
            name="password"
            id="sign-in_password-input"
            required
          />
        </label>
        <button id="sign-in_button" type="submit" onClick={handleSignIn}>
          SIGN IN
        </button>
      </form>
    </div>
  );
};

export default SignIn;
