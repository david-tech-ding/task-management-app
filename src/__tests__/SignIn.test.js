import React from "react";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import axios from "axios";
import SignIn from "../components/SignIn.js";
import { signInWithEmailAndPassword } from "firebase/auth";

jest.mock("../config/firebase", () => {
  return {
    auth: {
      currentUser: {
        updateProfile: jest.fn(),
      },
    },
  };
});

jest.mock("firebase/auth", () => {
  return {
    signInWithEmailAndPassword: jest.fn(() =>
      Promise.resolve({ user: { displayName: "testUser" } })
    ),
  };
});

jest.mock("axios", () => ({
  get: jest.fn(() => Promise.resolve({ data: [] })),
}));

describe("SignIn component", () => {
  const onSetLoggedInUser = jest.fn();
  const loggedInUser = { id: "", userName: "", firstName: "" };

  it("renders the component correctly", () => {
    const { asFragment } = render(
      <Router>
        <SignIn
          onSetLoggedInUser={onSetLoggedInUser}
          loggedInUser={loggedInUser}
        />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("updates state when input fields change", () => {
    render(
      <Router>
        <SignIn
          onSetLoggedInUser={onSetLoggedInUser}
          loggedInUser={loggedInUser}
        />
      </Router>
    );
    const emailInput = screen.getByLabelText("Email");
    const passwordInput = screen.getByLabelText("Password");

    fireEvent.change(emailInput, { target: { value: "test@example.com" } });
    fireEvent.change(passwordInput, { target: { value: "password123" } });

    expect(emailInput.value).toBe("test@example.com");
    expect(passwordInput.value).toBe("password123");
  });

  it("submits the form correctly", async () => {
    render(
      <Router>
        <SignIn
          onSetLoggedInUser={onSetLoggedInUser}
          loggedInUser={loggedInUser}
        />
      </Router>
    );
    const emailInput = screen.getByLabelText("Email");
    const passwordInput = screen.getByLabelText("Password");
    const submitButton = screen.getByText("SIGN IN");

    fireEvent.change(emailInput, { target: { value: "test@example.com" } });
    fireEvent.change(passwordInput, { target: { value: "password123" } });

    axios.get.mockResolvedValue({ data: [{}] });
    signInWithEmailAndPassword.mockResolvedValue({
      user: {
        email: "test@example.com",
        password: "password123",
        displayName: "testdisplayName",
      },
    });

    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(signInWithEmailAndPassword).toHaveBeenCalledWith(
        expect.any(Object),
        "test@example.com",
        "password123"
      );
    });
    await waitFor(() => {
      expect(axios.get).toHaveBeenCalledWith("/user/username/testdisplayName");
    });
  });
});
