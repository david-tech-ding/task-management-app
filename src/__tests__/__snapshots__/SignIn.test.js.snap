// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn component renders the component correctly 1`] = `
<DocumentFragment>
  <div
    class="auth"
  >
    <p
      aria-live="assertive"
      class="offscreen"
    />
    <h2
      class="auth-subheading"
    >
      Sign In
    </h2>
    <form
      class="auth-form"
    >
      <label
        for="sign-in_email-input"
      >
        Email
        <input
          class="auth-form_input"
          id="sign-in_email-input"
          name="email"
          required=""
          type="email"
          value=""
        />
      </label>
      <label
        for="sign-in_password-input"
      >
        Password
        <input
          class="auth-form_input"
          id="sign-in_password-input"
          name="password"
          required=""
          type="password"
          value=""
        />
      </label>
      <button
        id="sign-in_button"
        type="submit"
      >
        SIGN IN
      </button>
    </form>
  </div>
</DocumentFragment>
`;
